File: src/query/Parser.php
<?php

namespace Ething\Query;

// replace bool by boolean

/*

This class will parse a query string into a MongoDB query object used to filter a collection.


 
*/
	

class Parser {
	
	private $constants = array();
	private $fields = array();
	private $operators = array();
	
	private $fieldFallback = null;
	
	public function __construct(array $fields = array(), array $constants = array()){
		
		// set defaults operators
		$this->addOperator( array(
			
			new Operator('==', function($field,$value){
					return array(
						(string)$field => $value->is('date') ? new \MongoDB\BSON\UTCDateTime($value->getDate()->getTimestamp()*1000) : $value->getValue()
					);
				}),
			new Operator('!=', function($field,$value){
					return array(
						(string)$field => array( '$ne' => $value->is('date') ? new \MongoDB\BSON\UTCDateTime($value->getDate()->getTimestamp()*1000) : $value->getValue() )
					);
				}),
			new Operator('exists', function($field){
					return array(
						(string)$field => array( '$exists' => true )
					);
				}, '*', false),
			new Operator('is', function($field,$value){
					switch(strtolower($value->getValue())){
						case 'string':
							return array(
								(string)$field => array( '$type' => 2 )
							);
						case 'boolean':
						case 'bool':
							return array(
								(string)$field => array( '$type' => 8 )
							);
						case 'number':
							return array(
								'$or' => array(
									array( (string)$field => array( '$type' => 1 ) ),
									array( (string)$field => array( '$type' => 16 ) ),
									array( (string)$field => array( '$type' => 18 ) )
								)
							);
						case 'integer':
							return array(
								'$or' => array(
									array( (string)$field => array( '$type' => 16 ) ),
									array( (string)$field => array( '$type' => 18 ) )
								)
							);
						case 'float':
						case 'double':
							return array(
								(string)$field => array( '$type' => 1 )
							);
						case 'null':
							return array(
								(string)$field => array( '$type' => 10 )
							);
						default:
							throw new InvalidQueryException("unknown type '{$value}'",$stream);
					}
				}, '*', 'string'),
			// number or date
			new Operator('<', function($field,$value){
					return array(
						(string)$field => array( '$lt' => $value->is('date') ? new \MongoDB\BSON\UTCDateTime($value->getDate()->getTimestamp()*1000) : $value->getValue() )
					);
				}, array('integer','double','date')),
			new Operator('>', function($field,$value){
					return array(
						(string)$field => array( '$gt' => $value->is('date') ? new \MongoDB\BSON\UTCDateTime($value->getDate()->getTimestamp()*1000) : $value->getValue() )
					);
				}, array('integer','double','date')),
			// number only
			new Operator('<=', function($field,$value){
					return array(
						(string)$field => array( '$lte' => $value->getValue() )
					);
				}, array('integer','double')),
			new Operator('>=', function($field,$value){
					return array(
						(string)$field => array( '$gte' => $value->getValue() )
					);
				}, array('integer','double')),
			// string only
			new Operator('^=', function($field,$value){ // start with
					return array(
						(string)$field => array( '$regex' => '^'.$value->getValue() )
					);
				}, array('string')),
			new Operator('$=', function($field,$value){ // end with
					return array(
						(string)$field => array( '$regex' => $value->getValue().'$' )
					);
				}, array('string')),
			new Operator('*=', function($field,$value){ // containing
					return array(
						(string)$field => array( '$regex' => $value->getValue() )
					);
				}, array('string')),
			new Operator('~=', function($field,$value){ // containing word
					return array(
						(string)$field => array( '$regex' => '(^|\s)'.$value->getValue().'($|\s)' )
					);
				}, array('string'))
			
		 ));
		$this->addField($fields);
		$this->addConstant($constants);
	}
	
	
	public function addConstant($name, $value = null){
		
		if(is_array($name)){
			foreach($name as $key => $value)
				$this->constants[$key] = $value;
		}
		else 
			$this->constants[$name] = $value;
	}
	
	public function getConstant($name){
		return isset($name) ? ( isset($this->constants[$name]) ? $this->constants[$name] : null ) : array_values($this->constants);
	}
	
	public function addField($field){
		if(is_array($field)){
			foreach($field as $value)
				$this->addField($value);
		}
		else if($field instanceof Field)
			$this->fields[(string)$field] = $field;
	}
	
	public function getField($name){
		return isset($name) ? ( isset($this->fields[$name]) ? $this->fields[$name] : null ) : array_values($this->fields);
	}
	
	public function addOperator($operator){
		if(is_array($operator)){
			foreach($operator as $value)
				$this->addOperator($value);
		}
		else if($operator instanceof Operator)
			$this->operators[(string)$operator] = $operator;
	}
	
	public function getOperator($syntax){
		return isset($syntax) ? ( isset($this->operators[$syntax]) ? $this->operators[$syntax] : null ) : array_values($this->operators);
	}
	
	
	public function setFieldFallback($fallback){
		$this->fieldFallback = is_callable($fallback) ? $fallback : null;
	}
	
	/*
	* Entry point
	*/
	public function parse($query){
		$stream = new Stream($query);
		return $this->parseExpression($stream);
	}
	
	
	
	private function parseField(Stream &$stream){
		$stream->skipSpace();
		if($field = $stream->read('/[a-z0-9\.\-_]+/i')){
			if(array_key_exists($field, $this->fields)){
				return $this->fields[$field];
			}
			else{
				if($this->fieldFallback){
					$f = call_user_func($this->fieldFallback, $field);
					if($f instanceof Field)
						return $f;
				}
				throw new InvalidQueryException("unknown field '{$field}'",$stream);
			}
		}
		else
			throw new InvalidQueryException('expected a field',$stream);
	}
	
	
	private function parseOperator(Stream &$stream){
		$stream->skipSpace();
		if(($op = $stream->read('/[\=\!\^\>\<\$\*~]+/')) || ($op = $stream->readWord())){
			if(array_key_exists($op, $this->operators)){
				return $this->operators[$op];
			}
			else
				throw new InvalidQueryException("unknown operator '{$op}'",$stream);
		}
		else
			throw new InvalidQueryException("expected an operator",$stream);
	}
	
	private function parseValue(Stream &$stream){
		$stream->skipSpace();
		
		// try to get string (double quotes or simple quotes)
		if ( ($v = $stream->read('/^(?:\"([^\"]*(?:\"\"[^\"]*)*)\")/')) || ($v = $stream->read('/^(?:\'([^\']*(?:\'\'[^\']*)*)\')/')) ){
			// quoted string
			$r = substr($v, 1, -1); // remove the quotes
			if($v[0] === '"'){
				$r = \stripcslashes($r);
			}
			return new Value($r);
		}
		else {
			$v = $stream->read('/[a-z0-9\.\-_+]+/i');
			
			// number ? (integer or float)
			if(is_numeric($v)){
				return new Value($v + 0);
			}
			// boolean ?
			else if( preg_match('/^(true|false)$/i',$v) ){
				return new Value((bool) preg_match('/true/i',$v));
			}
			// maybe a CONSTANT
			else if(array_key_exists($v, $this->constants)){
				return new Value($this->constants[$v]);
			}
			else
				throw new InvalidQueryException("invalid value '{$v}'",$stream);
		}
		
	}
	
	/*
	 parse a Field-Operator-Value expression
	*/
	private function parseFOV(Stream &$stream){
		
		if($stream->read('/^\s*not($|\s+)/i')){
			return array( '$nor' => array($this->parseFOV($stream)) );
		}
		
		$field = $this->parseField($stream);
		$op = $this->parseOperator($stream);
		$value = $op->hasValue() ? $this->parseValue($stream) : new Value();
		
		if(!$op->accept($field, $value)){
			throw new InvalidQueryException("the operator '{$this}' is invalid");
		}
		
		try {
			$compiled = $field->compil($op,$value);
		}
		catch(\Exception $e){
			throw new InvalidQueryException($e->getMessage(),$stream);
		}
		
		return $compiled;
	}
	
	
	
	/*
	 * Parse a FOV or a enclosed expression
	*/
	private function parseFragment(Stream &$stream){
		if($stream->read('/^\s*\(/')){
			// start enclosure
			$f = $this->parseExpression($stream);
			
			// must be followed by a ')'
			if(!$stream->read('/^\s*\)/'))
				throw new InvalidQueryException("expected a ')'",$stream);
			
		}
		else
			$f = $this->parseFOV($stream);
			
		return $f;
	}
	
	private function parseExpression(Stream &$stream){
		
		$stack = array();
		
		$stream->skipSpace();
		
		if(!$stream->length() || $stream->match('/^\)/'))
			throw new InvalidQueryException('empty expression',$stream);
		
		// get the first fragment
		$stack[] = $this->parseFragment($stream);
		
		
		// search for other FOV expressions separated by a logical operator
		while($stream->length()){
			
			$stream->skipSpace();
			
			if(!$stream->length() || $stream->match('/^\)/'))
				break; // the end
			
			// search for a logical operator
			if($logic = $stream->read('/^(and|or)($|\s+)/i')){
				$stack[] = strtolower(trim($logic));
			}
			else
				throw new InvalidQueryException('waiting for a logical operator',$stream);
			
			// search for a fragment
			$stack[] = $this->parseFragment($stream);
			
		}
		
		//compilation of the stack
		$logicPrecedence = array('and','or');
		$logicIndex = 0;
		while(count($stack)>1){
			
			$currentLogic = $logicPrecedence[$logicIndex];
			
			for($i=0; $i<count($stack); $i++){
				if($stack[$i] === $currentLogic){
					$exp = array(
						'$'.$currentLogic => array( $stack[$i-1] , $stack[$i+1] )
					);
					
					array_splice ( $stack, $i-1, 3, array($exp) );
					
					$i--;
				}
			}
			
			$logicIndex++;
			
			if($logicIndex == count($logicPrecedence)){
				if(count($stack)>1)
					throw new InvalidQueryException('error');
				break;
			}
		}
		
		return $stack[0];
		
	}
	
}

File: ething/query/Parser.py
# coding: utf-8


# replace bool by boolean

"""

This class will parse a query string into a MongoDB query object used to filter a collection.


 
"""

import unicodedata
import re
import sys
from .Value import Value
from .Field import Field
from .Operator import Operator, EqualOperator, NotEqualOperator, ExistOperator, IsOperator, LowerOperator, GreaterOperator, LowerOrEqualOperator, GreaterOrEqualOperator, StartWithOperator, EndWithOperator, ContainOperator, ContainWordOperator, HasOperator
from .Stream import Stream
from .InvalidQueryException import InvalidQueryException
from future.utils import iteritems, listvalues
import ast


def get_type_from_json_schema(schema):

    type = schema.get('type')
    format = schema.get('format')

    if type is None:
        if 'anyOf' in schema or 'oneOf' in schema:
            types = []
            for s in schema['anyOf' if 'anyOf' in schema else 'oneOf']:
                t = get_type_from_json_schema(s)
                if isinstance(t, list):
                    types += t
                elif t == '*':
                    return t
                else:
                    types.append(t)
            return list(set(types))
        elif 'allOf' in schema:
            for s in schema['allOf']:
                t = get_type_from_json_schema(s)
                if t != '*':
                    return t

    if type == 'string':
        if format == "date-time":
            return "date"

    if type is None:
        return '*'

    return type


class Parser(object):

    def __init__(self, fields=None, constants=None, tz = None):

        self.tz = tz
        self.rules = []
        self.fields = {}
        self.constants = {}
        self.operators = {}
        self.fieldFallback = None

        # set defaults operators
        self.addOperator([
            EqualOperator(),
            NotEqualOperator(),
            ExistOperator(),
            IsOperator(),
            LowerOperator(),
            GreaterOperator(),
            LowerOrEqualOperator(),
            GreaterOrEqualOperator(),
            StartWithOperator(),
            EndWithOperator(),
            ContainOperator(),
            ContainWordOperator(),
            HasOperator(),
        ])

        self.addConstant('null', None)

        if fields:
            self.addField(fields)

        if constants:
            self.addConstant(constants)

    def addConstant(self, name, value=None):
        if isinstance(name, dict):
            for key, value in iteritems(name):
                self.constants[key] = value
        else:
            self.constants[name] = value

    def getConstant(self, name=None):
        if name is None:
            return listvalues(self.constants)
        if name in self.constants:
            return self.constants[name]
        return None

    def addField(self, field):
        if isinstance(field, list):
            for value in field:
                self.addField(value)
        else:
            self.fields[field.name] = field

    def getField(self, name):
        if name is None:
            return listvalues(self.fields)
        if name in self.fields:
            return self.fields[name]
        return None

    def addOperator(self, operator):
        if isinstance(operator, list):
            for value in operator:
                self.addOperator(value)
        else:
            self.operators[str(operator)] = operator

    def getOperator(self, syntax):
        if syntax is None:
            return listvalues(self.operators)
        if syntax in self.operators:
            return self.operators[syntax]
        return None

    def setFieldFallback(self, fallback):
        self.fieldFallback = fallback if callable(fallback) else None

    def add_field_rule(self, regexp, callback):
        self.rules.append({
            'regexp': regexp,
            're': re.compile(regexp),
            'callback': callback
        })

    def import_fields_from_json_schema(self, schema, name=None, helper=None):

        type = get_type_from_json_schema(schema)

        if type == 'object':

            properties = schema.get('properties', {})

            for prop_name in properties:
                self.import_fields_from_json_schema(
                    properties[prop_name], ("%s.%s" % (name, prop_name)) if name else prop_name)

            additionalProperties = schema.get('additionalProperties', True)

            if additionalProperties is not False:
                sub_type = None

                if isinstance(additionalProperties, dict):
                    sub_type = get_type_from_json_schema(additionalProperties)

                self.add_field_rule(
                    r'^%s.[^\.]+' % name, lambda field: Field(field, type=sub_type))

        if name:

            field = Field(name, type=type)

            if callable(helper):
                field = helper(name=name, type=type,
                               schema=schema, default_field=field)

            if isinstance(field, Field):
                self.addField(field)

    """
    Entry point
    """

    def parse(self, query):
        stream = Stream(query)
        return self.parseExpression(stream)

    def parseField(self, stream):
        stream.skipSpace()
        field = stream.read('[a-zA-Z0-9\.\-_]+')
        if field:
            if field in self.fields:
                return self.fields[field]
            else:

                for rule in self.rules:
                    if rule['re'].search(field):
                        f = rule['callback'](field)
                        if isinstance(f, Field):
                            return f

                if self.fieldFallback is not None:
                    f = self.fieldFallback(field)
                    if isinstance(f, Field):
                        return f
                raise InvalidQueryException(
                    "unknown field '%s'" % field, stream)
        else:
            raise InvalidQueryException('expected a field', stream)

    def parseOperator(self, stream):
        stream.skipSpace()
        op = stream.read('[\=\!\^\>\<\$\*~]+')
        if not op:
            op = stream.readWord()
        if op:
            if op in self.operators:
                return self.operators[op]
            else:
                raise InvalidQueryException(
                    "unknown operator '%s'" % op, stream)
        else:
            raise InvalidQueryException("expected an operator", stream)

    def parseValue(self, stream):
        stream.skipSpace()
        v = stream.read('^(?:\"([^\"]*(?:\"\"[^\"]*)*)\")')
        if not v:
            v = stream.read('^(?:\'([^\']*(?:\'\'[^\']*)*)\')')
        # try to get string (double quotes or simple quotes)
        if v:
            # quoted string

            return Value(self, ast.literal_eval(v))

            # r = v[1:-1] # remove the quotes
            # if(v[0] == '"'):
            #    r = r.decode('string_escape')
            #
            # return Value(self, r)

        else:
            v = stream.read('[a-zA-Z0-9\.\-_+]+')

            # number ? (integer or float)
            isNumber = False
            try:
                c = float(v)
                return Value(self, c)
            except ValueError:
                pass

            try:
                c = unicodedata.numeric(v)
                return Value(self, c)
            except (TypeError, ValueError):
                pass

            # boolean ?
            m = re.search('^True$', v, re.IGNORECASE)
            if m:
                return Value(self, True)
            m = re.search('^False$', v, re.IGNORECASE)
            if m:
                return Value(self, False)

            # maybe a CONSTANT
            if v in self.constants:
                return Value(self, self.constants[v])

            raise InvalidQueryException("invalid value '%s'" % v, stream)

    """
     parse a Field-Operator-Value expression
    """

    def parseFOV(self, stream):

        if stream.read('^\s*[nN][oO][tT]($|\s+)'):
            return {'$nor': [self.parseFOV(stream)]}

        field = self.parseField(stream)
        op = self.parseOperator(stream)
        value = self.parseValue(stream) if op.hasValue() else None

        if not op.accept(field, value):
            raise InvalidQueryException(
                "the operator '%s' is invalid" % op, stream)

        try:
            compiled = field.compil(op, value)
        except:
            raise InvalidQueryException(sys.exc_info()[1], stream)

        return compiled

    """
     Parse a FOV or a enclosed expression
    """

    def parseFragment(self, stream):
        if stream.read('^\s*\('):
            # start enclosure
            f = self.parseExpression(stream)

            # must be followed by a ')'
            if not stream.read('^\s*\)'):
                raise InvalidQueryException("expected a ')'", stream)

        else:
            f = self.parseFOV(stream)

        return f

    def parseExpression(self, stream):

        stack = []

        stream.skipSpace()

        if not stream.length() or stream.match('/^\)/'):
            raise InvalidQueryException('empty expression', stream)

        # get the first fragment
        stack.append(self.parseFragment(stream))

        # search for other FOV expressions separated by a logical operator
        while stream.length() > 0:

            stream.skipSpace()

            if not stream.length() or stream.match('^\)'):
                break  # the end

            # search for a logical operator
            logic = stream.read('^([aA][nN][dD]|[oO][rR])($|\s+)')
            if logic:
                stack.append(logic.strip().lower())
            else:
                raise InvalidQueryException(
                    'waiting for a logical operator', stream)

            # search for a fragment
            stack.append(self.parseFragment(stream))

        # compilation of the stack
        logicPrecedence = ['and', 'or']
        logicIndex = 0
        while len(stack) > 1:

            currentLogic = logicPrecedence[logicIndex]

            i = 0
            while i < len(stack):
                if stack[i] == currentLogic:
                    exp = {
                        '$'+currentLogic: [stack[i-1], stack[i+1]]
                    }

                    stack.pop(i+1)
                    stack.pop(i)
                    stack.pop(i-1)
                    stack.insert(i, exp)

                    i -= 1
                i += 1

            logicIndex += 1

            if logicIndex == len(logicPrecedence):
                if len(stack) > 1:
                    raise InvalidQueryException('error')
                break

        return stack[0]
